django/db/migrations/graph.py

```diff
@@ -32,26 +80,37 @@ class MigrationGraph(object):
     """
 
     def __init__(self):
+        self.node_map = {}
         self.nodes = {}
-        self.dependencies = {}
-        self.dependents = {}
+        self.cached = False
 
-    def add_node(self, node, implementation):
+    def add_node(self, key, implementation):
+        node = Node(key)
+        self.node_map[key] = node
         self.nodes[node] = implementation
```

Comment:
I think we can get rid of `self.nodes` when we add `implementation` to the node itself: `node = Node(key, implementation)`